### PR TITLE
Added a persistent log implementation

### PR DESCRIPTION
**Pull request by richorama**
which is not thread safe thanks to IEnumerable :¬(

Need to have a chat about concurrency..

### CODE REVIEW

File: Orleans.Consensus.Contract/Log/LogEntry.cs [hunk 1]
```diff
@@ -8,15 +8,16 @@ namespace Orleans.Consensus.Contract.Log
...
         {
             this.Id = entryId;
             this.Operation = operation;
         }
 
-        public TOperation Operation { get; }
+        public TOperation Operation { get; set; }
 
-        public LogEntryId Id { get; }
+        public LogEntryId Id { get; set; }
```
Comment (ReubenBond) [inline Orleans.Consensus.Contract/Log/LogEntry.cs]: Could we make these setters private? I am guessing you added them so protobuf can access them

**Comment by richorama** [comment]
Yeah, protobuf can't deserialize with private setters.

**Comment by ReubenBond** [comment]
I think it only accepts private setters if the property has the correct attributes, which we ought to do anyhow (otherwise we don't get forwards/backwards compatibility).

If we mark the class `[ProtoContract]` and the fields `[ProtoMember(1)]`, `[ProtoMember(2)]`, etc, it should work

**Comment by ReubenBond** [comment]
Fantastic work, by the way! I'm reviewing it now.

**Comment by richorama** [comment]
Yes. I went back and forth on this. I didn't like the abstraction leak of adding the attrbute, as the serializer can be swapped out. We also don't know what TOperation is, so can't decorate it...

File: Orleans.Consensus/Log/ProtobufSerializer.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    using ProtoBuf.Meta;
+    using System;
+    using System.IO;
+    using System.Threading.Tasks;
+
+    public class ProtobufSerializer<T> : ISerializer<T> where T:new()
+    {
+        RuntimeTypeModel model;
+
+        public ProtobufSerializer(params Type[] subTypes)
```
Comment (ReubenBond) [inline Orleans.Consensus/Log/ProtobufSerializer.cs:12]: I wonder if RuntimeTypeModel should be injected. Not a change we need to make before merging.

File: Orleans.Consensus/Log/StreamLog.cs [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+            }
+        }
+
+        void AddToCache(LogEntry<TOperation> entry)
+        {
+            this.cache.Add(entry.Id.Index, entry);
+            while (cache.Count > CACHE_SIZE)
+            {
+                // remove keys with lower indexes
+                cache.Remove(cache.Keys.Min());
```
Comment (ReubenBond) [inline Orleans.Consensus/Log/StreamLog.cs:56]: (Not blocking) This seems expensive: it looks like it visits up to 1000 entries for each add. Maybe we should use a different data structure for the cache.

**Comment by ReubenBond** [comment]
I'm happy to make consumers use protobuf-net, similarly for Autofac - too many abstractions often mean we can't use the unique features of a specific technology, such as forwards/backwards compatible serialization.

In this case, we can use `[DataMember(Order=1)]`, etc, attributes to achieve the same effect without leaking protobuf attributes.

File: Orleans.Consensus/Log/StreamLog.cs (hunk 3 above)
Comment (richorama) [inline Orleans.Consensus/Log/StreamLog.cs:56]: This while loop should only spin at most once for every add.

Comment (richorama) [inline Orleans.Consensus/Log/StreamLog.cs:56]: I agree that it's not a perfect fit though...

**Comment by richorama** [comment]
Ok. I'll play around with some attributes tomorrow...

File: Orleans.Consensus/Log/StreamLog.cs [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+        }
+
+        void BuildIndex()
+        {
+            stream.Position = 0;
+            var first = true;
+            while (stream.Position < stream.Length)
+            {
+                var position = stream.Position;
+                var entry = serializer.Deserialize(stream).Result;
```
Comment (ReubenBond) [inline Orleans.Consensus/Log/StreamLog.cs]: I'm not aware of any serializers which work in async, maybe we should just make it synchronous. Marc Gravell suggests async serialization is not worthwhile here: http://stackoverflow.com/a/11187593/635314, and that serializing to a buffer may be better (maybe RecyclableMemoryStream if we need it)

File: Orleans.Consensus/Log/StreamLog.cs [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+            if (!result) throw new ArgumentOutOfRangeException("cannot rewind stream to earlier position");
+            
+            // purge entries from the btree
+            for (var i = index; i <= this.LastLogEntryId.Index; i++)
+            {
+                bTreeIndex.Delete(i);
+                if (this.cache.ContainsKey(i)) this.cache.Remove(i);
+            }
+
+            // warning: this method does not update the LastLogEntryId
```
Comment (ReubenBond) [inline Orleans.Consensus/Log/StreamLog.cs:114]: Should this method also update the FirstLogEntryId? (I'm not sure yet)

File: Orleans.Consensus/Log/StreamLog.cs [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+            }
+        }
+
+        public IEnumerable<LogEntry<TOperation>> GetReverseCursor()
+        {
+            if (this.stream.Length == 0) yield break;
+
+            for (var i = this.LastLogEntryId.Index; i >= this.FirstLogEntryId.Index; i--)
+            {
+                yield return Get(i);
```
Comment (ReubenBond) [inline Orleans.Consensus/Log/StreamLog.cs:182]: If we use this a lot, maybe we should consider creating some kind of `ReverseEnumerator` class on the tree.

**Comment by ReubenBond** [comment]
Looks great :) I think testing is the only valid way to determine the correctness of the BTree implementation, but everything looks good.

**Comment by ReubenBond** [comment]
I'll leave merging to you, in case you've made some changes locally and want to squash or something.

**Comment by cmello** [comment]
Just watching your code and learning by now. Awesome work guys, congratulations!!

File: Orleans.Consensus/Log/StreamLog.cs (hunk 4 above)
Comment (richorama) [inline Orleans.Consensus/Log/StreamLog.cs]: the original motivation for making the serializer async was to maintain async all the way down. Other serializers might support async writing to the stream, alternatively the serializer could write to a memory stream, which could then write to the physical steam async (as you suggested).

In some recent performance testing I found that async = slow.

Comment (ReubenBond) [inline Orleans.Consensus/Log/StreamLog.cs]: Async can be slow, there are all sorts of nuances. It can be faster than blocking, though.

**Pull request merged**